java-checks/src/main/java/org/sonar/java/checks/PrivateFieldUsedLocallyCheck.java

```diff
@@ -0,0 +1,285 @@
+
+  private void initMethodsLiveFields(TypeSymbol classSymbol) {
+    fieldsLiveInMethodEntryPerMethod = new HashMap<>();
+    fieldsReadOnAnotherInstance = new HashSet<>();
+
+    for (Symbol memberSymbol : classSymbol.memberSymbols()) {
+
+      if (memberSymbol.isMethodSymbol() && memberSymbol.declaration() != null) {
+        MethodTree methodTree = (MethodTree) memberSymbol.declaration();
+
+        if (methodTree.block() != null) {
+          CFG cfg = CFG.build(methodTree);
+          LiveFields liveFields = LiveFields.analyze(cfg);
+          fieldsLiveInMethodEntryPerMethod.put(methodTree, liveFields.getIn(cfg.entry()));
+          fieldsReadOnAnotherInstance.addAll(liveFields.fieldsReadOnAnotherInstance);
+        }
+      }
+    }
+  }
+
+  private static boolean isPrivateField(Symbol memberSymbol) {
+    return memberSymbol.isPrivate() && memberSymbol.isVariableSymbol();
+  }
+
+  private static class LiveFields {
+
+    private final CFG cfg;
+    private final Map<Block, Set<Symbol>> out = new HashMap<>();
+    private Map<Block, Set<Symbol>> in;
+    private Set<Symbol> fieldsReadOnAnotherInstance = new HashSet<>();
+
+    private LiveFields(CFG cfg) {
+      this.cfg = cfg;
+    }
+
+    Set<Symbol> getIn(CFG.Block block) {
+      return in.get(block);
+    }
+
+    static LiveFields analyze(CFG cfg) {
+      LiveFields liveFields = new LiveFields(cfg);
+      final Map<CFG.Block, Set<Symbol>> in = new HashMap<>();
+      // Generate kill/gen for each block in isolation
+      Map<CFG.Block, Set<Symbol>> kill = new HashMap<>();
+      Map<CFG.Block, Set<Symbol>> gen = new HashMap<>();
+      for (CFG.Block block : liveFields.cfg.reversedBlocks()) {
+        liveFields.fieldsReadOnAnotherInstance.addAll(fieldsReadOnAnotherInstance(block));
+        Set<Symbol> blockKill = new HashSet<>();
+        Set<Symbol> blockGen = new HashSet<>();
+        liveFields.processBlockElements(block, blockKill, blockGen);
+        kill.put(block, blockKill);
+        gen.put(block, blockGen);
+      }
+      liveFields.analyzeCFG(in, kill, gen);
+      // out of exit block are empty by definition.
+      if (!liveFields.out.get(liveFields.cfg.reversedBlocks().get(0)).isEmpty()) {
+        throw new IllegalStateException("Out of exit block should be empty");
+      }
+
+      liveFields.in = in;
+      return liveFields;
+    }
+
+    private static Set<Symbol> fieldsReadOnAnotherInstance(Block block) {
+      Set<Symbol> fieldsReadOnAnotherInstance = new HashSet<>();
+      for (Tree element : block.elements()) {
+        if (element.is(Kind.MEMBER_SELECT)) {
+          MemberSelectExpressionTree memberSelect = (MemberSelectExpressionTree) element;
+          if (memberSelect.expression().is(Kind.IDENTIFIER)) {
+            String objectName = ((IdentifierTree) memberSelect.expression()).name();
+            Symbol symbol = memberSelect.identifier().symbol();
+
+            if (!symbol.isStatic() && !"this".equals(objectName) && isField(symbol)) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS134)